frontend/util.lua

```diff
@@ -1635,4 +1635,18 @@ function util.round_decimal(num, points)
     return math.floor(num * op) / op
 end
 
+function util.isExecutable(file)
+    local attrs = lfs.attributes(file)
+    return attrs and attrs.mode == "file" and os.execute("test -x " .. file) == 0
```

Comment:
> No, this is meaningless 

Emphasis on _something like_, of course.

Reply:
But using the shell is no biggie if it'd take a lot of extra work. ;-)

Reply:
> we have access to [access]

inside joke: we also had `command` command :P (I am starting to admire these repetitions :D)

Reply:
OK, so the terminal plugin does not work on my Nexus 7 tablet when using the default detected shell (bash). After switch to `sh`, I can run `bash` manually however:
![2024-11-21-193744_690x1104_scrot](https://github.com/user-attachments/assets/2c3d8207-f335-420e-926d-f9250a5e2087)

So I'm not sure what's happening (and I've never used it before). ¯\\_(ツ)_/¯

Base PR for access bits + `isExecutable` moved to `ffi.util`: https://github.com/koreader/koreader-base/pull/1982

Diff for this PR (fixing `util.which` along the way :P):

```diff
 frontend/util.lua                  | 8 ++------
 plugins/terminal.koplugin/main.lua | 5 +++--
 2 files changed, 5 insertions(+), 8 deletions(-)

diff --git i/frontend/util.lua w/frontend/util.lua
index 8675285bc..8e1a1b686 100644
--- i/frontend/util.lua
+++ w/frontend/util.lua
@@ -1635,15 +1635,11 @@ function util.round_decimal(num, points)
     return math.floor(num * op) / op
 end
 
-function util.isExecutable(file)
-    local attrs = lfs.attributes(file)
-    return attrs and attrs.mode == "file" and os.execute("test -x " .. file) == 0
-end
-
 function util.which(command, path)
     path = path or os.getenv("PATH") or ""
     for p in path:gmatch("([^:]+)") do
-        if util.isExecutable(p .. "/" .. command) then
+        p = p .. "/" .. command
+        if ffiUtil.isExecutable(p) then
             return p
         end
     end
diff --git i/plugins/terminal.koplugin/main.lua w/plugins/terminal.koplugin/main.lua
index cab84eea6..5deb56c4f 100644
--- i/plugins/terminal.koplugin/main.lua
+++ w/plugins/terminal.koplugin/main.lua
@@ -80,7 +80,8 @@ local bit = require("bit")
 local lfs = require("libs/libkoreader-lfs")
 local _ = require("gettext")
 local C_ = _.pgettext
-local T = require("ffi/util").template
+local ffiUtil = require("ffi/util")
+local T = ffiUtil.template
 
 local CHUNK_SIZE = 80 * 40 -- max. nb of read bytes (reduce this, if taps are not detected)
 
@@ -95,7 +96,7 @@ local Terminal = WidgetContainer:extend{
 
 function Terminal:isExecutable(file)
     -- check if file is an executable or a command in PATH
-    return util.isExecutable(file) or util.which(file)
+    return ffiUtil.isExecutable(file) or util.which(file)
 end
 
 -- Try SHELL environment variable and some standard shells
```

Reply:
Thanks!
I am currently working on GH from web browser, I'll try to make these diffs carefully by hand :D

Reply:
Done.
Can someone please bump base?
Or please show me show to do it from the browser?

Reply:
@benoit-pierre 
> getcwd: cannot access parent directories

Do you think this is because Android, and you have no access to `/bin`, where from `bash` is trying to start? 
https://stackoverflow.com/questions/29396928/error-retrieving-current-directory-getcwd-cannot-access-parent-directories

Reply:
Also, https://github.com/Homebrew/brew/issues/9118#issuecomment-726460964

Reply:
Bash is actually in `/system/xbin/bash` on that system. But yeah, I don't know.